### PR TITLE
[FEATURE] [needs-docs] Add circular data dependencies

### PR DESCRIPTION
**Pull request by troopa81**
## Description

This PR is the first step of the proposed GRANT [Snap Cache Improvements](https://docs.google.com/document/d/1VzgLJ-hy2GtcWrAy5geivhDDdXRGJUDcikCWUtXRe9k/edit#heading=h.uau3di2m4nvn). 

If fixes the QGEP issue described [here](https://github.com/QGEP/QGEP/issues/492)

It add the possibility to define circular data dependencies (layer1 depends on layer2 that depends on layer 1), and by extension on the layer itself (layer1 depends on layer1). This last is useful for instance when you modify one line in a line layer that trigger connected lines modifications on database side. In this case, `dataChanged` signal needs to be fired in order to refresh the snapping cache (and avoid what is described in QGEP issue). 

This PR doesn't propose any optimisations on snap cache indexing. It will be done in a separate PR.

**API BREAK** : I remove the `QgsMapLayer::hasDependencyCycle` method because it has no use anymore. Even if the method was protected, it changes the API. Is it an issue? Is it better to keep it and mark it as deprecated? I remove it for now.

cc @mhugo @lbartoletti @m-kuhn @haubourg 

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Review by mhugo — APPROVED** [review]
Good for me. It seems to suppress more code than it adds, even better :)

File: src/core/qgsmaplayer.h [hunk 1]
```diff
@@ -1499,9 +1499,6 @@ class CORE_EXPORT QgsMapLayer : public QObject
     //! List of layers that may modify this layer on modification
     QSet<QgsMapLayerDependency> mDependencies;
 
-    //! Checks whether a new set of dependencies will introduce a cycle
-    bool hasDependencyCycle( const QSet<QgsMapLayerDependency> &layers ) const;
```
Comment (nyalldawson) [inline src/core/qgsmaplayer.h]: Instead of removing this and breaking API, I'd suggest leaving it, marking it as deprecated, and always returning false

Comment (troopa81) [reply src/core/qgsmaplayer.h]: OK 8983695

**Comment by troopa81** [comment]
Please don't stale!

**Comment by haubourg** [comment]
just tested in production environment and it works nicely! 
fixed_snapping_index_out_of_sync_for_transactiongroup_and_dbtriggers

@nyalldawson or @m-kuhn  Can someone merge this. I think having this in LTR will be a great relief for many application using database side intelligence :)

Of course this slows the digitizing work, but #31374  should help.

**Comment by haubourg** [comment]
Note for documentation team (note for myself), this feature is barely visible in the UI. 
It only adds one line in the layer depency dialog.

**Comment by Chaz6** [comment]
Thanks for this improvement! I would really like to read a blog post about using this feature and its applications (i.e. what kind of editing tasks with some simple sql examples).

**Comment by troopa81** [comment]
@Chaz6 Here is a [blog post](https://oslandia.com/en/2020/01/10/en-qgis-snapping-improvements/) where we talk (among other things) about data depedencies

**Pull request merged**